README.md
# SOFI
**S**l**o**t**F**illing and **I**ntents

[![SOFI](https://img.shields.io/badge/Name-SOFI-ff69b4.svg)](https://img.shields.io/badge/Language-Python3-ff69b4.svg)
[![Language](https://img.shields.io/badge/Language-Python3-ff69b4.svg)](https://img.shields.io/badge/Language-Python3-ff69b4.svg)

##
[Embeddings-Tutorial](https://blog.keras.io/using-pre-trained-word-embeddings-in-a-keras-model.html)

## Installation

## Usage

## Contributing
- Fabian Fey
- Gianna Weber

## License



changes.txt
Changes to bert:

run_classifier.py
Line 316: return ["0", "1"] ==> return ["1", "2", ...]
Line 323: if i == 0: ==> if i == 1:
Line 329: label = "0" ==> label = "1"
Line 375: return ["0", "1"] ==> return ["1", "2", ...]

postagtest.py
# https://nlpforhackers.io/lstm-pos-tagger-keras/

import numpy as np
import logzero

from sklearn.model_selection import train_test_split
from keras import preprocessing
from keras.preprocessing.sequence import pad_sequences

from keras.models import Sequential
from keras.layers import Dense, LSTM, InputLayer, Bidirectional, TimeDistributed, Embedding, Activation
from keras.optimizers import Adam

logzero.logfile("Log/intent-logs.log", maxBytes=1e6, backupCount=5)
logzero.loglevel(10)
logger = logzero.logger

LEARNING_RATE = 0.001
BATCH_SIZE = 128
EPOCHS = 20


def main ():
    utterances = loadUtterances()
    utterance_slots = loadSlots() 
    
    ##  Splits the utterances and slots into train- an test-sets
    (train_utterances, 
    test_utterances, 
    train_slots, 
    test_slots) = train_test_split(utterances, utterance_slots, test_size=0.33, random_state=42)

    ##  Create an index dictionary for utterances and slots
    words_set, slots_set = set([]), set([]) 
    
    for utterance in train_utterances:
        for word in utterance:
            words_set.add(word)
    
    for slots in train_slots:
        for slot in slots:
            slots_set.add(slot)
    
    word2index = {word: i + 2 for i, word in enumerate(list(words_set))}
    word2index['-PAD-'] = 0  # The special value used for padding
    word2index['-OOV-'] = 1  # The special value used for OOVs
    
    slot2index = {slot: i + 1 for i, slot in enumerate(list(slots_set))}
    slot2index['-PAD-'] = 0  # The special value used to padding

    ##  Convert utterances and slots to numbers
    train_utterances_X, test_utterances_X, train_slots_Y, test_slots_Y = [], [], [], []
    
    for utterance in train_utterances:
        utterance_int = [] # utterance with indices
        for word in utterance:
            try:
                utterance_int.append(word2index[word])
            except KeyError:
                utterance_int.append(word2index['-OOV-'])
    
        train_utterances_X.append(utterance_int)
    
    for utterance in test_utterances:
        utterance_int = []
        for word in utterance:
            try:
                utterance_int.append(word2index[word])
            except KeyError:
                utterance_int.append(word2index['-OOV-'])
    
        test_utterances_X.append(utterance_int)
    
    for utterance in train_slots:
        train_slots_Y.append([slot2index[slot] for slot in utterance])
    
    for utterance in test_slots:
        test_slots_Y.append([slot2index[slot] for slot in utterance])

    logger.debug("Train Utterance: %s", train_utterances_X[0])
    logger.debug("Test Utterance: %s", test_utterances_X[0])
    logger.debug("Train Slots: %s", train_slots_Y[0])
    logger.debug("Test Slots: %s", test_slots_Y[0])

    ## Find max sentence length for padding
    MAX_LENGTH = len(max(train_utterances_X, key=len))
    logger.debug("Maximum utterance lenght: %s", MAX_LENGTH)

    ## Pad the utterances and slots (keras)
    train_utterances_X = pad_sequences(train_utterances_X, maxlen=MAX_LENGTH, padding='post')
    test_utterances_X = pad_sequences(test_utterances_X, maxlen=MAX_LENGTH, padding='post')
    train_slots_Y = pad_sequences(train_slots_Y, maxlen=MAX_LENGTH, padding='post')
    test_slots_Y = pad_sequences(test_slots_Y, maxlen=MAX_LENGTH, padding='post')
    
    logger.debug("Train Utterance: %s", train_utterances_X[0])
    logger.debug("Test Utterance: %s", test_utterances_X[0])
    logger.debug("Train Slots: %s", train_slots_Y[0])
    logger.debug("Test Slots: %s", test_slots_Y[0])

    ## Define the model
    model = Sequential()
    model.add(InputLayer(input_shape=(MAX_LENGTH, )))
    model.add(Embedding(len(word2index), 128, mask_zero=True))
    model.add(Bidirectional(LSTM(256, return_sequences=True)))
    model.add(TimeDistributed(Dense(len(slot2index))))
    model.add(Activation('softmax'))
    
    model.compile(loss='categorical_crossentropy',
                optimizer=Adam(LEARNING_RATE),
                metrics=['accuracy'])
    
    model.summary()

    cat_train_slots_y = to_categorical(train_slots_Y, len(slot2index))
    logger.debug("Slots to categorical: %s", cat_train_slots_y[0])

    cat_test_slots_y = to_categorical(test_slots_Y, len(slot2index))

    ## Train
    model.fit(train_utterances_X, cat_train_slots_y, batch_size=BATCH_SIZE, epochs=EPOCHS)#, validation_split=0.2)

    ## Validate
    scores = model.evaluate(test_utterances_X, cat_test_slots_y)
    print(f"{model.metrics_names[1]}: {scores[1] * 100}")   # acc: 99.09751977804825

 
## One-Hot-Enc for tags
def to_categorical(sequences, categories):
    cat_sequences = []
    for utterance in sequences:
        cats = []
        for item in utterance:
            cats.append(np.zeros(categories))
            cats[-1][item] = 1.0
        cat_sequences.append(cats)
    return np.array(cat_sequences)

def loadSlots():
    with open("TrainingData/Slots/train_slot_label.tsv", 'r') as devLabelFile:
        slotList = []
        for line in devLabelFile:
            line = line.replace("_", "")
            line = line.replace("\n", "")
            line = line.split("\t")
            slotList.append(line)
    return slotList

def loadUtterances():
    with open("TrainingData/Slots/train_slot_Utt.tsv", 'r') as devUttFile:
        utterances = []
        for utterance in devUttFile:
            utterancePreProc = preprocessing.text.text_to_word_sequence(utterance, filters='?!,.:;\n', lower=False, split='\t')
            utterances.append(utterancePreProc)
    return utterances

if __name__ == "__main__":
    main()
    

preprocessing.py
#!/usr/bin/env python3
"""
intent-preprocessing.py

Preprocesses the data

*full.json => train.tsv
validate_*.json => test.tsv
*.json => dev.tsv

1= BookRestaurant
2 = SearchScreeningEvent
3 = RateBook
4 = PlayMusic
5 = AddToPlaylist
6 = GetWeather
7 = SearchCreativeWork

@Author: Fabian Fey, Gianna Weber
"""

import os
import json
import codecs

path_train=[]
path_dev=[]
path_test=[]
intent_dict = {"BookRestaurant":1,
"SearchScreeningEvent":2,
"RateBook":3,
"PlayMusic":4,
"AddToPlaylist":5,
"GetWeather":6,
"SearchCreativeWork":7
}


def removeChar(str, n):
    first_part = str[:n] 
    last_part = str[n+1:]
    return first_part + last_part


for subdir, dirs, files in os.walk('Data/'):
    for file in files:
        if "train" in file and "full" not in file:  # dev
            path_dev.append(os.path.join(subdir, file))
        elif "full" in file:  # train
            path_train.append(os.path.join(subdir, file))
        elif "validate" in file:  # test
            path_test.append(os.path.join(subdir, file))


def intents(paths, filename):
    with open(filename, 'w') as OutputFile:
        for path in paths:
            with open(path, 'r', encoding='ISO-8859-1') as jsonFile:
                print(path)
                file = json.load(jsonFile)
                for intent in file:  # {"AddToPlaylist":[...]}
                    #print(intent)
                    for data in file[intent]:  # {"AddToPlaylist":[{"data": [...]}, {...}, ...]}
                        #print(data)
                        for key in data:
                            # print(key)
                            text = []
                            for entry in data[key]:
                                #print(entry["text"])
                                tmpText = entry["text"] # phrases for particular slots
                                tmpText = tmpText.replace("\n", "")
                                tmpText = tmpText.replace("  ", " ") # to circumvent newlines withing utterance/phrase, e.g. " \n "
                                text.append(tmpText) # total utterance 
                            if len("".join(text)) != 0:
                                print(str(intent_dict[intent])+"\t"+"".join(text)) # mapping from 1-7 to possible intents
                                OutputFile.write(str(intent_dict[intent])+"\t"+"".join(text)+"\n") # create new data files (dev, test, train) for intents


def slots(paths, slotLabel, slotUtterance):
    with open(slotLabel, 'w') as SlotLabelOutputFile:
        with open(slotUtterance, 'w') as SlotUtteranceOutputFile:
            for path in paths: # iterate over all files
                with open(path, 'r', encoding='ISO-8859-1') as jsonFile: 
                    file = json.load(jsonFile)
                    for intent in file:  # {"AddToPlaylist":[...]}
                        # print(intent)
                        for data in file[intent]:  # {"AddToPlaylist":[{"data": [...]}, {...}, ...]}
                            # print(data)
                            for key in data:
                                # print(key)
                                text = []
                                entity = []
                                for entry in data[key]:
                                    entryText = []
                                    if "entity" in entry: # slot "entity"
                                        utterance = entry["text"] # part of utterances 

                                        utterance = utterance.replace("\n", "")
                                        utterance = utterance.replace("  ", " ")
                                        textSplit = utterance.split(" ") # split at whitespace
                                        
                                        # "text": "Cita Romántica" has "entity": "playlist" -> text = ["Cita", "Romántica"], entity = ["playlist", "playlist"]
                                        for i in range(0, len(textSplit)): 
                                            entity.append(entry["entity"])
                                        text.extend(textSplit)
                                        # print("Text: %s", text)
                                    else:
                                        utterance = entry["text"]
                                        utterance = utterance.replace("\n", "")
                                        utterance = utterance.replace("  ", " ")
                                        
                                        # Problem with preceding and following whitespaces: needed to be removed
                                        if utterance == " ":
                                            continue
                                        elif utterance[:1] == " ":
                                            if utterance[-1:] == " ":
                                                tmp = removeChar(utterance, 0)
                                                tmp2 = removeChar(tmp, len(tmp)-1)
                                                entryText.extend(tmp2.split(" "))
                                            else:
                                                tmp = removeChar(utterance, 0)
                                                entryText.extend(tmp.split(" "))
                                        elif utterance[-1:] == " ":
                                            tmp = removeChar(utterance, len(utterance)-1)
                                            entryText.extend(tmp.split(" "))
                                        else:
                                            entryText.extend(utterance.split(" "))

                                        for i in range(0, len(entryText)):
                                            entity.append("NULL")
                                    
                                    text.extend(entryText)

                                #print("".join(text))
                                utterance = "\t".join(text)
                                labels = "\t".join(entity)
                                
                                utterance = utterance.replace("\n", "")
                                utterance = utterance.replace("  ", " ")
                                labels = labels.replace("\n", "")
                                
                                SlotLabelOutputFile.write(labels + "\n")
                                SlotUtteranceOutputFile.write(utterance + "\n") 

# Will be removed in a future version
def slotCount(path_dev, path_test, path_train):
    paths = []
    paths.extend(path_dev)
    paths.extend(path_test)
    paths.extend(path_train)
    SlotDict = {}
    for path in paths:
        with open(path, 'r', encoding='ISO-8859-1') as jsonFile:
            file = json.load(jsonFile)
            for intent in file:  # {"AddToPlaylist":[...]}
                # print(intent)
                for data in file[intent]:  # {"AddToPlaylist":[{"data": [...]}, {...}, ...]}
                    # print(data)
                    for key in data:
                        # print(key)
                        text = []
                        entity = []
                        for entry in data[key]:
                            entryText = []
                            if "entity" in entry:
                                slot = entry["entity"]
                                slot = slot.replace("_", "")
                                if slot in SlotDict:
                                    SlotDict[slot] +=1
                                else:
                                    SlotDict[slot] = 1
                            else:
                                if "NULL" in SlotDict:
                                    SlotDict["NULL"] +=1
                                else:
                                    SlotDict["NULL"] = 1

    with open("SlotList.tsv", "w") as slotFile:
        for key in SlotDict:
            slotFile.write(str(key+"\t"))
            print(key, SlotDict[key]) # Slot Verteilung
        print(len(SlotDict)) # 39 entity, i.e. slot, types

       
#slotCount(path_dev, path_test, path_train)
slots(path_dev, "dev_slot_label.tsv", "dev_slot_Utt.tsv")
slots(path_test, "test_slot_label.tsv", "test_slot_Utt.tsv")
slots(path_train, "train_slot_label.tsv", "train_slot_Utt.tsv")
# intents(path_dev, "dev.tsv")
# intents(path_test, "test.tsv")
# intents(path_train, "train.tsv")

slot-classifier.py
#!/usr/bin/env python3
"""
intent-classifier.py

Embedding Layer:
https://blog.cambridgespark.com/tutorial-build-your-own-embedding-and-use-it-in-a-neural-network-e9cde4a81296
https://github.com/CyberZHG/keras-bert

@Author: Fabian Fey, Gianna Weber
"""


import numpy as np
import logging
import logzero
import pickle
import os
# import msgpack
# import msgpack_numpy as msgp
# msgp.patch()

import tensorflow.keras as keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation, Embedding, Flatten, LSTM, SpatialDropout1D
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

from bert_embedding import BertEmbedding

from sklearn.model_selection import train_test_split

DATA_SET_TYPE = "dev"

MAX_SENTENCE_LENGTH = 30
EMBEDDING_SIZE = 768

EPOCHS = 50
BATCH_SIZE = 30
ACTIVATION = ""
OPTIMIZER = 'adam'
LOSS = 'categorical_crossentropy'
DROPOUT = 0.4

logzero.logfile("Log/slot-logs.log", maxBytes=1e6, backupCount=5)
logzero.loglevel(10)
logger = logzero.logger

def main():
    logger.info("----Start of Program----")

    bert_embedding = BertEmbedding()

    sentences_Train, sentences_Test, labels_Train, labels_Test, vocab_size, tokenizerSentences = process_Data()

    logger.warning("vocab_size: %s", vocab_size)

    # X_train, Y_train = load_Dev()
   
    # if os.path.isfile("Cache/slotEmbeddingMatrix.pickle"):
    if os.path.isfile("Cache/slotEmbeddingMatrix.pickle"):
        logger.info("Loading existing embedding matrix")
        with open("Cache/slotEmbeddingMatrix.pickle", "rb") as file:
            embedding_matrix = pickle.load(file)
            # embedding_matrix = msgpack.unpackb(file, default=msgp.encode)
        logger.info("Finished Loading Embedding Matrix")

    else: 
        logger.info("Creating Embedding Matrix")
        # create a weight matrix for words in training utterancePreProc
        embedding_matrix = np.zeros((vocab_size, EMBEDDING_SIZE))

        for word, i in tokenizerSentences.word_index.items():
            # Creating a list on which to put the word since the python implementation of BERT want's a list not a single word
            tmpList = []
            tmpList.append(word)
            #logger.info("Search Embedding: %s", word)
            embedding_vector = bert_embedding(tmpList)
            #logger.info("Found Embedding")
            # print(embedding_vector[0][1][0])
            if embedding_vector is not None:
                try:
                    embedding_matrix[i] = embedding_vector[0][1][0]
                except:
                    logger.exception("i: %s", i)

        embedding_matrix = np.insert(embedding_matrix, 0, np.zeros(EMBEDDING_SIZE), axis=0)

        print(embedding_matrix)

        with open("Cache/slotEmbeddingMatrix.pickle", "wb") as file:
            pickle.dump(embedding_matrix, file)
            # packed = msgpack.packb(embedding_matrix, default=msgp.encode)
            # file.write(packed)

        logger.info("Finished Embedding Matrix Creation, dumped it.")
    # define model
    model = Sequential()
    e = Embedding(vocab_size + 1, EMBEDDING_SIZE, input_length=MAX_SENTENCE_LENGTH, weights=[embedding_matrix], trainable=False, mask_zero=True)
    model.add(e)
    model.add(SpatialDropout1D(DROPOUT))
    model.add(LSTM(64, dropout=DROPOUT, recurrent_dropout=DROPOUT))
    model.add(Dense(7))
    model.add(Activation("softmax"))
    #model.add(Dense(7, activation='softmax'))

    # compile the model
    logger.info("Compiling model")
    model.compile(optimizer=OPTIMIZER, loss=LOSS, metrics=['acc'])
    # summarize the model
    logger.info("Model Summary: %s", str(model.summary()))
    # fit the model
    logger.info("Fitting model")
    model.fit(sentences_Train, labels_Train, epochs=EPOCHS, verbose=0)
    # evaluate the model
    logger.info("Evaluating model")
    loss, accuracy = model.evaluate(sentences_Test, labels_Test, verbose=0)
    logger.info('Accuracy: %f' % (accuracy*100))
    logger.info("----End of Programm----")


def load_Data(dataset):
    tokenizerSlots = Tokenizer()

    sentences = []
    slotIndices = []
    slotList = LoadSlotList()

    tokenizerSlots.fit_on_texts(slotList)

    if dataset == "dev":
        logger.info("Loading 'dev' Data")
        with open("TrainingData/Slots/dev_slot_label.tsv", 'r') as devLabelFile:
            for line in devLabelFile:
                line = line.split("\t")
                sequence = tokenizerSlots.texts_to_sequences(line)
                slotIndices.extend(sequence)

        with open("TrainingData/Slots/dev_slot_Utt.tsv", 'r') as devUttFile:
            for utterance in devUttFile:

                utterancePreProc = keras.preprocessing.text.text_to_word_sequence(utterance, filters='?!,.:;\n', lower=False, split=' ')
                sentences.append(utterancePreProc)

    elif dataset == "test":
        logger.info("Loading 'test' Data")

        with open("TrainingData/Slots/test_slot_label.tsv", 'r') as devLabelFile:
            for line in devLabelFile:
                line = line.split("\t")
                sequence = tokenizerSlots.texts_to_sequences(line)
                slotIndices.extend(sequence)

        with open("TrainingData/Slots/test_slot_Utt.tsv", 'r') as devUttFile:
            for utterance in devUttFile:

                utterancePreProc = keras.preprocessing.text.text_to_word_sequence(utterance, filters='?!,.:;\n', lower=False, split=' ')
                sentences.append(utterancePreProc)

    else:
        logger.info("Loading 'train' Data")
        with open("TrainingData/Slots/train_slot_label.tsv", 'r') as devLabelFile:
            for line in devLabelFile:
                line = line.split("\t")
                sequence = tokenizerSlots.texts_to_sequences(line)
                slotIndices.extend(sequence)

        with open("TrainingData/Slots/train_slot_Utt.tsv", 'r') as devUttFile:
            for utterance in devUttFile:

                utterancePreProc = keras.preprocessing.text.text_to_word_sequence(utterance, filters='?!,.:;\n', lower=False, split=' ')
                sentences.append(utterancePreProc)
        
    return sentences, slotIndices

def LoadSlotList():

    with open("TrainingData/Slots/SlotList.tsv", "r") as LoadSlotFile:
        slotList = []
        for line in LoadSlotFile:
            lineSplit = line.split("\t")
            for item in lineSplit:
                if item != "":
                    slotList.append(item)

        # logger.debug("all slots: %s", slotList)
        # logger.debug("Lenght SlotList: %s", len(slotList))
    
    return(slotList)

def process_Data():
    tokenizerSentences = Tokenizer()

    allSentences = []

    sentencesTrain, goldTrain = load_Data("train")
    sentencesTest, goldTest = load_Data("test")
    sentencesDev, goldDev = load_Data("dev")

    logger.warning("sentences: len train: %s, len test: %s, len dev: %s", len(sentencesTrain), len(sentencesTest), len(sentencesDev))
    logger.warning("gold: len train: %s, len test: %s, len dev: %s", len(goldTrain), len(goldTest), len(goldDev))
    
    allSentences.extend(sentencesDev)
    allSentences.extend(sentencesTest)
    allSentences.extend(sentencesTrain)

    logger.warning("len all sentences: %s", len(allSentences))
    # logger.warning("Sentences Train: %s", sentencesTrain)
    tokenizerSentences.fit_on_texts(allSentences)
    vocab_size = len(tokenizerSentences.word_index)  # + 2  # define the amount of tokens, +1 b/c of zero vector, ?+1 unknown?

    if DATA_SET_TYPE == "dev":
        logger.info("Splitting 'dev' Data")
        sentences_Train, sentences_Test, labels_Train, labels_Test = train_test_split(sentencesDev, goldDev, test_size=0.33, random_state=42)

        # Create an int encoded version of each token
        X_Train_Encoded = tokenizerSentences.texts_to_sequences(sentences_Train)
        X_Test_Encoded = tokenizerSentences.texts_to_sequences(sentences_Test)

        # Pad each sentence to a length of MAX_SENTENCE_LENGTH
        X_Train_Padded = pad_sequences(X_Train_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')
        X_Test_Padded = pad_sequences(X_Test_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')

        return X_Train_Padded, X_Test_Padded, labels_Train, labels_Test, vocab_size, tokenizerSentences

    elif DATA_SET_TYPE == "test":
        logger.info("Splitting 'test' Data")
        sentences_Train, sentences_Test, labels_Train, labels_Test = train_test_split(sentencesTest, goldTest, test_size=0.33, random_state=42)

        # Create an int encoded version of each token
        X_Train_Encoded = tokenizerSentences.texts_to_sequences(sentences_Train)
        X_Test_Encoded = tokenizerSentences.texts_to_sequences(sentences_Test)

        # Pad each sentence to a length of MAX_SENTENCE_LENGTH
        X_Train_Padded = pad_sequences(X_Train_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')
        X_Test_Padded = pad_sequences(X_Test_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')

        return X_Train_Padded, X_Test_Padded, labels_Train, labels_Test, vocab_size, tokenizerSentences

    else:
        logger.info("Splitting 'train' Data")
        sentences_Train, sentences_Test, labels_Train, labels_Test = train_test_split(sentencesTrain, goldTrain, test_size=0.33, random_state=42)

        # Create an int encoded version of each token
        X_Train_Encoded = tokenizerSentences.texts_to_sequences(sentences_Train)
        X_Test_Encoded = tokenizerSentences.texts_to_sequences(sentences_Test)

        # Pad each sentence to a length of MAX_SENTENCE_LENGTH
        X_Train_Padded = pad_sequences(X_Train_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')
        X_Test_Padded = pad_sequences(X_Test_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')

        return X_Train_Padded, X_Test_Padded, labels_Train, labels_Test, vocab_size, tokenizerSentences


if __name__ == "__main__":
    main()


intent-classifier.py
#!/usr/bin/env python3
"""
intent-classifier.py

Embedding Layer:
https://blog.cambridgespark.com/tutorial-build-your-own-embedding-and-use-it-in-a-neural-network-e9cde4a81296
https://github.com/CyberZHG/keras-bert

@Author: Fabian Fey, Gianna Weber
"""


import numpy as np
import logging
import logzero
import pickle
import os

import tensorflow.keras as keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation, Embedding, Flatten, LSTM, SpatialDropout1D
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

from bert_embedding import BertEmbedding

from sklearn.model_selection import train_test_split

DATA_SET_TYPE = "dev"

MAX_SENTENCE_LENGTH = 30
EMBEDDING_SIZE = 768 # lenght of vector created by BERT

EPOCHS = 50
BATCH_SIZE = 30
ACTIVATION = ""
OPTIMIZER = 'adam'
LOSS = 'categorical_crossentropy'
DROPOUT = 0.4

logzero.logfile("Log/intent-logs.log", maxBytes=1e6, backupCount=5)
logzero.loglevel(10)
logger = logzero.logger

def main():
    logger.info("----Start of Program----")

    bert_embedding = BertEmbedding()

    sentences_Train, sentences_Test, labels_Train, labels_Test, vocab_size, tokenizer = process_Data()

    logger.debug("vocab_size: %s", vocab_size)

    if os.path.isfile("Cache/embeddingMatrix.pickle"): # Check for embedding matrix
        logger.info("Loading existing embedding matrix")
        with open("Cache/embeddingMatrix.pickle", "rb") as file:
            embedding_matrix = pickle.load(file)
        logger.info("Finished Loading Embedding Matrix")

    else: 
        logger.info("Creating Embedding Matrix")
        # create a weight matrix for words in training
        embedding_matrix = np.zeros((vocab_size, EMBEDDING_SIZE)) # create matrix with zeros

        for word, i in tokenizer.word_index.items(): # check index per word
            # Creating a list on which to put the word since the python implementation of BERT want's a list, not a single word
            tmpList = []
            tmpList.append(word)
            embedding_vector = bert_embedding(tmpList)

            if embedding_vector is not None:
                try:
                    embedding_matrix[i] = embedding_vector[0][1][0]
                except:
                    logger.exception("i: %s", i)

        embedding_matrix = np.insert(embedding_matrix, 0, np.zeros(EMBEDDING_SIZE), axis=0) # 0 = OOV, but not needed

        print(embedding_matrix)

        with open("Cache/embeddingMatrix.pickle", "wb") as file:
            pickle.dump(embedding_matrix, file)

        logger.info("Finished Embedding Matrix Creation, dumped it.")
    # define model
    model = Sequential()
    e = Embedding(vocab_size + 1, EMBEDDING_SIZE, input_length=MAX_SENTENCE_LENGTH, weights=[embedding_matrix], trainable=False, mask_zero=True)
    model.add(e)
    model.add(SpatialDropout1D(DROPOUT))
    model.add(LSTM(64, dropout=DROPOUT, recurrent_dropout=DROPOUT))
    model.add(Dense(7))
    model.add(Activation("softmax"))

    # compile the model
    logger.info("Compiling model")
    model.compile(optimizer=OPTIMIZER, loss=LOSS, metrics=['acc'])
    # summarize the model
    logger.info("Model Summary: %s", str(model.summary()))
    # fit the model
    logger.info("Fitting model")
    model.fit(sentences_Train, labels_Train, epochs=EPOCHS, verbose=0)
    # evaluate the model
    logger.info("Evaluating model")
    loss, accuracy = model.evaluate(sentences_Test, labels_Test, verbose=0)
    logger.info('Accuracy: %f' % (accuracy*100))
    logger.info("----End of Programm----")


def load_Data(dataset):
    utterances = []
    one_hot_intents = []

    if dataset == "dev":
        logger.info("Loading 'dev' Data")
        with open("TrainingData/Intents/dev.tsv", 'r') as devFile:
            for line in devFile:
                line = line.split("\t")
                
                utterance = line[1] # [0] = intent, [1] = utterance
                
                utterancePreProc = keras.preprocessing.text.text_to_word_sequence(utterance, filters='?!,.:;\n', lower=False, split=' ') # tokenizer, returns each utterance as list
                utterances.append(utterancePreProc)

                if int(line[0]) == 1:
                    encoded = np.array([1, 0, 0, 0, 0, 0, 0])
                elif int(line[0]) == 2:
                    encoded = np.array([0, 1, 0, 0, 0, 0, 0])
                elif int(line[0]) == 3:
                    encoded = np.array([0, 0, 1, 0, 0, 0, 0])
                elif int(line[0]) == 4:
                    encoded = np.array([0, 0, 0, 1, 0, 0, 0])
                elif int(line[0]) == 5:
                    encoded = np.array([0, 0, 0, 0, 1, 0, 0])
                elif int(line[0]) == 6:
                    encoded = np.array([0, 0, 0, 0, 0, 1, 0])
                elif int(line[0]) == 7:
                    encoded = np.array([0, 0, 0, 0, 0, 0, 1])

                one_hot_intents.append(encoded)
                npArray = np.array(one_hot_intents)
    elif dataset == "test":
        logger.info("Loading 'test' Data")
        with open("TrainingData/Intents/test.tsv", 'r') as devFile:
            for line in devFile:
                line = line.split("\t")
                
                utterance = line[1]
                
                utterancePreProc = keras.preprocessing.text.text_to_word_sequence(utterance, filters='?!,.:;\n', lower=False, split=' ')
                utterances.append(utterancePreProc)

                if int(line[0]) == 1:
                    encoded = np.array([1, 0, 0, 0, 0, 0, 0])
                elif int(line[0]) == 2:
                    encoded = np.array([0, 1, 0, 0, 0, 0, 0])
                elif int(line[0]) == 3:
                    encoded = np.array([0, 0, 1, 0, 0, 0, 0])
                elif int(line[0]) == 4:
                    encoded = np.array([0, 0, 0, 1, 0, 0, 0])
                elif int(line[0]) == 5:
                    encoded = np.array([0, 0, 0, 0, 1, 0, 0])
                elif int(line[0]) == 6:
                    encoded = np.array([0, 0, 0, 0, 0, 1, 0])
                elif int(line[0]) == 7:
                    encoded = np.array([0, 0, 0, 0, 0, 0, 1])

                one_hot_intents.append(encoded)
                npArray = np.array(one_hot_intents)
    else:
        logger.info("Loading 'train' Data")
        with open("TrainingData/Intents/train.tsv", 'r') as trainFile:
            for line in trainFile:
                line = line.split("\t")
                
                utterance = line[1]
                
                utterancePreProc = keras.preprocessing.text.text_to_word_sequence(utterance, filters='?!,.:;\n', lower=False, split=' ')
                utterances.append(utterancePreProc)

                if int(line[0]) == 1:
                    encoded = np.array([1, 0, 0, 0, 0, 0, 0])
                elif int(line[0]) == 2:
                    encoded = np.array([0, 1, 0, 0, 0, 0, 0])
                elif int(line[0]) == 3:
                    encoded = np.array([0, 0, 1, 0, 0, 0, 0])
                elif int(line[0]) == 4:
                    encoded = np.array([0, 0, 0, 1, 0, 0, 0])
                elif int(line[0]) == 5:
                    encoded = np.array([0, 0, 0, 0, 1, 0, 0])
                elif int(line[0]) == 6:
                    encoded = np.array([0, 0, 0, 0, 0, 1, 0])
                elif int(line[0]) == 7:
                    encoded = np.array([0, 0, 0, 0, 0, 0, 1])

                one_hot_intents.append(encoded)
                npArray = np.array(one_hot_intents)
                
    return utterances, npArray


def process_Data():
    tokenizer = Tokenizer()

    allSentences = []

    sentencesTrain, goldTrain = load_Data("train")
    sentencesTest, goldTest = load_Data("test")
    sentencesDev, goldDev = load_Data("dev")

    logger.warning("utterances: len train: %s, len test: %s, len dev: %s", len(sentencesTrain), len(sentencesTest), len(sentencesDev))
    logger.warning("gold: len train: %s, len test: %s, len dev: %s", len(goldTrain), len(goldTest), len(goldDev))

    allSentences.extend(sentencesDev)
    allSentences.extend(sentencesTest)
    allSentences.extend(sentencesTrain)

    logger.warning("len all utterances: %s", len(allSentences))

    tokenizer.fit_on_texts(allSentences)
    vocab_size = len(tokenizer.word_index)  # + 3  # define the amount of tokens, +1 b/c of zero vector, ?+1 unknown?

    if DATA_SET_TYPE == "dev":
        logger.info("Splitting 'dev' Data")
        sentences_Train, sentences_Test, labels_Train, labels_Test = train_test_split(sentencesDev, goldDev, test_size=0.33, random_state=42)

        # Create an int encoded version of each token
        X_Train_Encoded = tokenizer.texts_to_sequences(sentences_Train)
        X_Test_Encoded = tokenizer.texts_to_sequences(sentences_Test)

        # Pad each sentence to a length of MAX_SENTENCE_LENGTH
        X_Train_Padded = pad_sequences(X_Train_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')
        X_Test_Padded = pad_sequences(X_Test_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')

        return X_Train_Padded, X_Test_Padded, labels_Train, labels_Test, vocab_size, tokenizer

    elif DATA_SET_TYPE == "test":
        logger.info("Splitting 'test' Data")
        sentences_Train, sentences_Test, labels_Train, labels_Test = train_test_split(sentencesTest, goldTest, test_size=0.33, random_state=42)

        # Create an int encoded version of each token
        X_Train_Encoded = tokenizer.texts_to_sequences(sentences_Train)
        X_Test_Encoded = tokenizer.texts_to_sequences(sentences_Test)

        # Pad each sentence to a length of MAX_SENTENCE_LENGTH
        X_Train_Padded = pad_sequences(X_Train_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')
        X_Test_Padded = pad_sequences(X_Test_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')

        return X_Train_Padded, X_Test_Padded, labels_Train, labels_Test, vocab_size, tokenizer

    else:
        logger.info("Splitting 'train' Data")
        sentences_Train, sentences_Test, labels_Train, labels_Test = train_test_split(sentencesTrain, goldTrain, test_size=0.33, random_state=42)

        # Create an int encoded version of each token
        X_Train_Encoded = tokenizer.texts_to_sequences(sentences_Train)
        X_Test_Encoded = tokenizer.texts_to_sequences(sentences_Test)

        # Pad each sentence to a length of MAX_SENTENCE_LENGTH
        X_Train_Padded = pad_sequences(X_Train_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')
        X_Test_Padded = pad_sequences(X_Test_Encoded, maxlen=MAX_SENTENCE_LENGTH, padding='post')

        return X_Train_Padded, X_Test_Padded, labels_Train, labels_Test, vocab_size, tokenizer


if __name__ == "__main__":
    main()
